### ISSUE TITLE
Add motion support

### ISSUE THREAD

**Issue by nefarius** [ViGEm/DsHidMini#217]
User @DestructoSphere has provided [extensions to support motion](https://github.com/DestructoSphere/DsHidMini/commit/62a024d95e77ac37d3f360cb90422effc43a8035) (many thanks!)

After the new config system is merged (#197) I'll start experimenting with this.

**Comment by mrwallace888** [comment]
I'm going to assume this G-axis issue is the reason why Sackboy has scoliosis in LittleBigPlanet?

At least the rest of his motion controls work fine.

**Comment by nefarius** [comment]
Research pass is far enough along to implement against. No driver changes yet; notes and dumps are on `feat/217-motion-research`.

**Write-up:** [docs/MOTION.md](https://github.com/nefarius/DsHidMini/blob/feat/217-motion-research/docs/MOTION.md) (status + checklist at the bottom). Tools/dumps: [research/ds3-motion/](https://github.com/nefarius/DsHidMini/tree/feat/217-motion-research/research/ds3-motion).

**What we now know** (checked against Sony's signed `sixaxis.sys`, eight physical pads, and the CircumSpector PS3 captures):

- Accel is a single gain of **113** over `(zero - oneG)` from EEPROM page `0xA0`. X is mirrored *after* calibration (`0x3FF - cal`), not on the raw value. Formula matches every live pose we took.
- Gyro sign: raw **falls** for clockwise-from-above; every `sixaxis.sys` path inverts it, so clockwise is **positive** in Sony's reported value. Scale is about **1.4 counts per deg/s** (~+/-360 deg/s full scale). The old "113 vs 226" question is closed (no 2x gain).
- Hardware cal byte is **26.4 counts/step** (Sony's constant, confirmed on an A2). Placement is `[5]/[6]` if Feature `0x01` lists field `0x07`, else `[3]/[4]` on the SIXAXIS path.
- Three gyro paths come from the Feature `0x01` **field list**, not the type bytes. One pad reports DS3-looking `18 18 18 18` but takes the SIXAXIS path. Sending the factory cal byte once is not enough on hardware-trim pads (A2 idles ~160 counts off until the tracker steps).
- Counterfeits return a plausible EEPROM template, so a `zero == oneG` check never fires. Detection has to be behavioural (frozen sensors / 920 Hz spam).

**What DsHidMini does wrong today (SIXAXIS.SYS mode):** gyro sign is backwards *and* unzeroed; X is mirrored on the raw value; no EEPROM read and no cal byte. DS4Windows mode still has no motion.

**Still open:** absolute gyro scale (needs a turntable, not more hand turns), whether pads without field `0x07` ignore the cal byte, a live SIXAXIS stream, `0xEF` over Bluetooth, DS4 axis permutation.

Next step is the driver work in `DsHid.c` / USB start, not more reverse engineering. Happy to take more pad dumps if anyone has an odd revision.

**Comment by nefarius** [comment]
Driver work for the USB path is in `master` via #531 (small viewer follow-up in #532).

**Write-up:** [docs/MOTION.md](https://github.com/nefarius/DsHidMini/blob/master/docs/MOTION.md) (implemented behaviour + deferred list at the bottom).

**Shipped**

- USB `SET`/`GET Feature 0xEF` page `0xA0` after F2/F5. A failed or invalid EEPROM read keeps nominal `zero=512`, `oneG=399` and does **not** fail device start.
- Per-report Sony processing: accel gain 113, X mirrored *after* cal, gyro path from the Feature `0x01` field list (`PLAIN_ZERO` / `HW_CAL` / `SIXAXIS`). The 49-byte SIXAXIS GetFeature now carries the calibrated host-order values. The 12-byte SIXAXIS input report is unchanged (no motion fields).
- Gyro tracker + cal-byte resend on `HW_CAL` / `SIXAXIS` USB pads (factory byte once is not enough on hardware-trim pads).
- Backward-compatible third IPC region + `GetMotionSnapshot`. The existing 60-byte raw HID slot is untouched; older SDKs still map the two older regions.
- ControlApp Motion viewer: numeric readout plus a diagnostic 3D pose (gravity pitch/roll, integrated yaw). Yaw will drift; use Recenter after holding the pad still.

Bluetooth uses the same formulas with the nominal fallback and does **not** attempt `0xEF` or send hardware cal bytes this pass. DS4Windows mode still leaves DS4 gyro/accel offsets 13-24 at zero.

**Still deferred**

- Bluetooth `0xEF` over the BthPS3 HID control channel
- DS4Windows motion-field mapping (axis permutation still needs a reference DS4 capture)
- Counterfeit freeze detection (both seen fakes return a plausible `0200 0180` EEPROM template)
- Turntable gyro scale (still ~1.4 counts/(deg/s) from hand turns)

The live USB checklist in MOTION.md still wants a `PLAIN_ZERO` and a `HW_CAL` pad. Extra dumps welcome, especially Bluetooth `0xEF` and a reference DS4 capture.

**Comment by nefarius** [comment]
Not bad!

https://github.com/user-attachments/assets/b5b3e2f2-5e07-4404-8bd0-ca40da0a098b

**Comment by nefarius** [comment]
Follow-up on live USB `HW_CAL` pads (CECHZC2E-A1 / DS3-A1b class) after #531. Both are in `master`.

**Viewer pose (#533).** Gravity pitch/roll were swapped against the USB/trigger edge, and integrated yaw was bound to the Helix roll axis, so residual gyro looked like a constant leftward roll. Pitch is USB/trigger, roll is the grips, yaw spins about the face normal. Recenter still only zeros yaw. The viewer also ignores rest rates below a few deg/s so EEPROM leftover does not creep the 3D pad.

**`HW_CAL` publish path (#535).** `sixaxis.sys` reports `clamp(0x3FF - raw)` and leaves residuals smaller than one cal-byte step (~26.4 counts) in the HID report. On the A1 that is a rock-steady ~16 counts / ~10.7 deg/s clockwise yaw — the tracker learns `zeroRef` but will not step `0x77`. DsHidMini now publishes the tracker output `clamp(512 + zeroRef - raw)` on IPC and the 49-byte SIXAXIS GetFeature when the tracker is initialized, so that bias does not reach consumers. Failed/invalid Feature `0xEF` still falls back to `0x3FF - raw`. The cal-byte tracker still runs for large trim errors (A2-class).

Write-up: [docs/MOTION.md](https://github.com/nefarius/DsHidMini/blob/master/docs/MOTION.md).

Needs a rebuilt **driver** for #535 (ControlApp-only is not enough). Extra `HW_CAL` / `PLAIN_ZERO` USB checks still welcome.

**Comment by nefarius** [comment]
Follow-up after live USB checks on an Obigben clone and a genuine 2E A1 (`PLAIN_ZERO`). Both are in `master`.

**Clone rest drift (#538).** Feature `0x01` clone-heuristic `PLAIN_ZERO` pads (Obigben, Fake DS3) seed Sony's gyro tracker in **software-only** mode: `zeroRef` follows rest, the cal byte is never stepped or sent. Genuine `PLAIN_ZERO` / `HW_CAL` / `SIXAXIS` paths are unchanged. The Obigben gyro is frozen (~500 vs template EEPROM zero 512), so this only removes the false ~8.6 deg/s clockwise rest; yaw will not follow turns. Needs a rebuilt **driver**.

**Viewer on-grip pose (#539).** Rolling a genuine pad onto a grip made the 3D model flip about yaw and left heading offset after returning flat. Display used Euler pitch from `atan2(ay, -az)` (undefined when gravity is along X) and integrated the face-normal gyro as heading while that axis was horizontal. The viewer now uses a shortest-arc tilt quaternion and weights yaw by how level the pad is (`-u_z`). Driver tracker unchanged.

A 2E A1 CSV showed no `ZeroRef` step on either grip. Rest `GyroMilliDps` is about -1 deg/s on this `PLAIN_ZERO` path: the deadzone hid it at rest, but it rode along during handling. The viewer now learns that rate on a clean flat rest, freezes it, and subtracts it only from samples that already pass the deadzone.

The Motion viewer Record button writes `%ProgramData%\DsHidMini\Log\Motion\motion-slot<N>-<timestamp>.csv` (one row per `SampleIndex`). Pause no longer mixes new sensor columns with a frozen pose.

Write-up: [docs/MOTION.md](https://github.com/nefarius/DsHidMini/blob/master/docs/MOTION.md).

ControlApp-only is enough for #539; #538 needs the new driver. Still deferred: Bluetooth `0xEF`, DS4Windows motion fields, counterfeit freeze detection, turntable gyro scale.

**Comment by nefarius** [comment]
Bluetooth motion follow-up is in `master` via #549 (which replaces the wireless GET path from #547).

**Write-up:** [docs/MOTION.md](https://github.com/nefarius/DsHidMini/blob/master/docs/MOTION.md) (Bluetooth cache + remaining checklist). Why there is no wireless EEPROM read: [docs/PS3_USB_STARTUP.md](https://github.com/nefarius/DsHidMini/blob/master/docs/PS3_USB_STARTUP.md#bluetooth-has-no-equivalent-read).

**What we got wrong in #547.** The first Bluetooth pass issued HIDP Feature GET for `0x01` / `0xEF` over BthPS3, reasoning from a file named `..._Bluetooth_Pairing.txt`. That capture — like every other one cited in `PS3_USB_STARTUP.md` — is a **USB** link-layer pairing trace. There is no Bluetooth capture of those reads in this repo. Linux `hid-sony`, the USB Host Shield `PS3BT` library, and the `OpenPuck` emulator also never `GET_REPORT` Feature `0x01` / `0xEF` over the air; the console only reads them on the cable while pairing. On a live genuine 2E A1 that path never loaded EEPROM (`Fallback` stayed set) and the raw gyro sat at a near-zero constant (~5). The tracker formula `clamp(512 + zero - raw)` turns that into a full-scale yaw spin regardless of calibration.

**Shipped (#549)**

- USB `SET` / `GET Feature 0xEF` page `0xA0` still runs as before, and now persists the 64-byte reply (plus identification) on the USB devnode. A failed or invalid read still keeps nominal `zero=512`, `oneG=399` and **clears** any stale cache so Bluetooth cannot pick up a corrupt blob.
- Bluetooth never queries the pad for `0x01` / `0xEF`. It matches the pad MAC against USB instances (including phantom / unplugged) and loads the cached blobs. A pad never seen over USB stays nominal, with no extra startup delay.
- ControlApp Motion viewer reports the source: live USB, cached from USB, or “connect this controller over USB once…”. Missing source properties no longer abort the viewer.
- Gated experimental `SET Feature 0xF4` on Bluetooth connect for pads with a cached, non-clone identification. Current payload is the USB value `42 0C 00 00`; Linux / USB-Host-Shield's Bluetooth `42 03 00 00` is in the tree to compare. This is the working hypothesis for the raw-gyro-5 symptom.

Needs a rebuilt **driver** (ControlApp-only is not enough for the cache or `0xF4`). Pair the pad over USB once after installing, then reconnect wirelessly.

**Still deferred / pending hardware**

- Confirming a cache hit on a live pad (identification + EEPROM match the USB session, including with the USB instance unplugged) and which `0xF4` payload — `42 0C`, `42 03`, or neither — brings Bluetooth gyro off the near-zero constant
- DS4Windows motion-field mapping (axis permutation still needs a reference DS4 capture)
- Counterfeit freeze detection
- Turntable gyro scale (still ~1.4 counts/(deg/s) from hand turns)

Extra Bluetooth dumps welcome, especially raw gyro before/after `0xF4` and a pad that has never been plugged in over USB on this machine.

**Comment by nefarius** [comment]
Live check of the #549 cache path on a genuine 2E A1 (`PLAIN_ZERO`, firmware `04 00 08`), driver 3.9.0.2278.

**Result:** works. USB once, then Bluetooth: identification and EEPROM load from the cache, `MotionCalibrationSource` reads `CachedFromUsb`, viewer shows factory calibration, and the wireless yaw spin from the #547 build is gone.

Devnode properties after the run:

- USB (phantom, unplugged): key 13 = 64-byte `0xEF` page `0xA0` reply `00 EF 04 00 08 03 01 A0 … 01 F2 01 82 | 01 EE 01 7F | 01 F0 01 83 | 01 E1 00 00`, key 14 = 1 (live USB)
- Bluetooth: key 4 identical to the USB copy, key 14 = 2 (cached from USB)

Decoded: accel X 498/386, Y 494/383, Z 496/387, gyro zero 481, cal byte 0. Matches the earlier USB session for this pad.

**One trap worth knowing.** First attempt looked like a cache miss: the `connect over USB once` hint stayed up after a USB round-trip and the pad kept rotating wirelessly. Both devnodes were missing keys 13/14 entirely - not even the `None` marker a failed read leaves - and the System log showed BthPS3 tearing down and re-creating the pad's PDO at the exact second the driver package was installed, while the pad was connected. The on-disk `dshidmini.dll` was 3.9.0, but WUDFHost was still running the previous image. A reboot fixed it with no code change. If you update the driver with a pad connected and Bluetooth still reports nominal calibration afterwards, reboot before filing a bug.

**Still open**

- `0xF4`: the shipped `42 0C 00 00` went out on this connect and motion is correct, but I have not isolated whether it was required or whether `42 03 00 00` behaves the same. Needs an A/B on the same pad.
- `HW_CAL` / `SIXAXIS` pads over Bluetooth (cal byte via the unified output payload) - untested wirelessly so far.
- DS4Windows motion fields, counterfeit freeze detection, turntable gyro scale - unchanged.
